service/sharddistributor/executorclient/client.go

```diff
@@ -0,0 +1,118 @@
+}
+
+type ShardProcessorFactory[SP ShardProcessor] interface {
+	NewShardProcessor(shardID string) (SP, error)
+}
+
+type Executor[SP ShardProcessor] interface {
+	Start(ctx context.Context)
+	Stop()
+
+	GetShardProcess(shardID string) (SP, error)
+}
+
+type Params[SP ShardProcessor] struct {
+	fx.In
+
+	Dispatcher            *yarpc.Dispatcher
+	DC                    *dynamicconfig.Collection
+	MetricsClient         metrics.Client
+	Logger                log.Logger
+	ShardProcessorFactory ShardProcessorFactory[SP]
+	Config                Config
+	TimeSource            clock.TimeSource
+}
+
+func NewExecutor[SP ShardProcessor](params Params[SP]) (Executor[SP], error) {
+	shardDistributorClient, err := createShardDistributorExecutorClient(params.Dispatcher, params.DC, params.MetricsClient, params.Logger)
+	if err != nil {
+		return nil, fmt.Errorf("create shard distributor executor client: %w", err)
+	}
+
+	// TODO: get executor ID from environment
+	executorID := uuid.New().String()
+
+	return &executorImpl[SP]{
+		logger:                 params.Logger,
+		shardDistributorClient: shardDistributorClient,
+		shardProcessorFactory:  params.ShardProcessorFactory,
+		heartBeatInterval:      params.Config.HeartBeatInterval,
+		namespace:              params.Config.Namespace,
+		executorID:             executorID,
+		timeSource:             params.TimeSource,
+		stopC:                  make(chan struct{}),
+	}, nil
+}
+
+func createShardDistributorExecutorClient(dispatcher *yarpc.Dispatcher, dc *dynamicconfig.Collection, metricsClient metrics.Client, logger log.Logger) (sharddistributorexecutor.Client, error) {
+	shardDistributorClientConfig, ok := dispatcher.OutboundConfig(service.ShardDistributor)
+	if !ok {
+		return nil, fmt.Errorf("no outbound config for shard distributor")
+	}
+	if !rpc.IsGRPCOutbound(shardDistributorClientConfig) {
+		return nil, fmt.Errorf("shard distributor client does not support non-GRPC outbound")
+	}
+
+	shardDistributorExecutorClient := grpc.NewShardDistributorExecutorClient(
+		sharddistributorv1.NewShardDistributorExecutorAPIYARPCClient(shardDistributorClientConfig),
+	)
+
+	shardDistributorExecutorClient = timeoutwrapper.NewShardDistributorExecutorClient(shardDistributorExecutorClient, timeoutwrapper.ShardDistributorExecutorDefaultTimeout)
+	if errorRate := dc.GetFloat64Property(dynamicproperties.ShardDistributorExecutorErrorInjectionRate)(); errorRate != 0 {
+		shardDistributorExecutorClient = errorinjectors.NewShardDistributorExecutorClient(shardDistributorExecutorClient, errorRate, logger)
+	}
+	if metricsClient != nil {
+		shardDistributorExecutorClient = metered.NewShardDistributorExecutorClient(shardDistributorExecutorClient, metricsClient)
+	}
+
+	return shardDistributorExecutorClient, nil
+}
+
+func Module[SP ShardProcessor]() fx.Option {
+	return fx.Module("shard-distributor-executor-client",
+		fx.Provide(NewExecutor[SP]),
+		fx.Invoke(func(executor Executor[SP], lc fx.Lifecycle) {
+			lc.Append(fx.Hook{
```

Comment:
nit: fx.StartStopHook(executor.Start, executor.Stop)
It is a generic wrapper that supports hooks without error